[DO NOT MERGE] build: avoid magic when cross-compiling

benesch opened the pull request:
Handing off to @tamird. Did a hasty rebase, so not positive this actually produces working release binaries. (It did at one point over the weekend, though.)

---

Avoid deriving variables from the target triple to reduce complexity.
This makes it slightly more painful to cross-compile, because more
variables must be specified manually. For our purposes, this is easily
solved by a new script, build/release.go, which invokes Make with
the appropriate cross-compilation options for our builder image's
toolchains.

benesch committed:
build: avoid magic when cross-compiling

Avoid deriving variables from the target triple to reduce complexity.
This makes it slightly more painful to cross-compile, because more
variables must be specified manually. For our purposes, this is easily
solved by a Go command, build/cmd/crdb-build-release, which invokes
Make with the appropriate cross-compilation options for our builder
image's toolchains.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15337)
<!-- Reviewable:end -->

benesch commented on build/makewrapper/main.go:
```diff
@@ -81,79 +79,75 @@ type releaseConfiguration struct {
+	}
+	for _, name := range []string{"GOFLAGS", "TAGS", "SUFFIX"} {
+		if value, ok := os.LookupEnv(name); ok {
+			args = append(args, name+"="+value)
```
Doesn't this break `SUFFIX`? (I tried to do the same thing initially.)

tamird replied:
Ah, yeah. SUFFIX needs special handling. Fixed.

benesch commented on build/makewrapper/main.go:
```diff
@@ -69,9 +69,7 @@ var releaseConfigurations = map[string]releaseConfiguration{
 
-var xtoolsBase = "/x-tools"
-
-var defaultTarget = "build"
```
whoop, thanks!

benesch committed:
build: avoid "built for newer OSX version" warnings

The default MACOSX_DEPLOYMENT_TARGET uses only the major and minor
version numbers, not the patch number. The linker will emit a warning
if e.g. we try to link a library built for 10.12.4 while the
MACOSX_DEPLOYMENT_TARGET is 10.12. This happens when you have used Make
to build the native dependencies, but run a naked `go build` or `go
test`.

Avoid triggering these warnings by setting the
MACOSX_DEPLOYMENT_TARGET to MAJOR.MINOR instead of MAJOR.MINOR.PATCH in
the Makefile.

tamird commented:
Haven't looked at the main command here yet, apologies. Will finish the review this afternoon, but I'm concerned about the direction this is going with respect to baking the building binary into the docker image.

---

Reviewed 6 of 6 files at r3, 7 of 12 files at r4.
Review status: 7 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[Makefile, line 60 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_jwEgpcBsHh07xUaw:-Kk_jwEgpcBsHh07xUax:b-c6ypik) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/Makefile#L60)):*
> ```Makefile
> endif
> 
> # We intentionally use LINKFLAGS instead of the more traditional LDFLAGS
> ```

this stuff used to be higher up in this file; is there a reason to relocate it? not relocating it has the benefit of a smaller diff and less git blame acquisition

---

*[Makefile, line 65 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_jul_xScyz3SeWjsT:-Kk_jul_xScyz3SeWjsU:b-7lkna3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/Makefile#L65)):*
> ```Makefile
> LINKFLAGS ?=
> 
> # TYPE=release builds a release build, which omit debug information to produce a
> ```

omits

---

*[build/Dockerfile, line 137 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kdFBT8wDSAMVRk9a:-Kk_kdFBT8wDSAMVRk9b:b-fw5e6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/Dockerfile#L137)):*
> ```Dockerfile
> 
> COPY cmd /root/go/src/cmd
> RUN GOBIN=/usr/local/bin go install ./src/cmd/... && rm -rf ./src/cmd
> ```

oh boy, can we please not do this? this means we need to rev the docker image every time we change this command. I understand that this command relies on the image, but I'd rather we make the command invoke the builder rather than the other way around.

---

*[build/teamcity-acceptance.sh, line 15 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kXKh87fD7dGKoS4I:-Kk_kXKh87fD7dGKoS4J:b-fo4x2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/teamcity-acceptance.sh#L15)):*
> crdb-build-release

do we need to install this first?

---

*[build/teamcity-build-test-binary.sh, line 5 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kUvSOJlasTdCDvZu:-Kk_kUvSOJlasTdCDvZv:b-fo4x2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/teamcity-build-test-binary.sh#L5)):*
> crdb-build-release

do we need to install this first?

---

*[build/teamcity-publish-artifacts.sh, line 18 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kRrdY_BIIJl7Vo1L:-Kk_kRrdY_BIIJl7Vo1M:b-fo4x2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/teamcity-publish-artifacts.sh#L18)):*
> crdb-build-release

do we need to install this first?

---

*[pkg/acceptance/prepare.sh, line 11 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kN50mt0bzJHVZhcP:-Kk_kN50mt0bzJHVZhcQ:b72uayv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/pkg/acceptance/prepare.sh#L11)):*
> crdb-build-release

do we need to `go install` this thing first?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-:-Kk_ku7y59IMzud8NjVk:bxvxdqq)*
<!-- Sent from Reviewable.io -->

benesch commented:
Ha, I baked it into the binary because I thought you had suggested that way
back when. That's like the least consequential part of this PR, fortunately.

N

On Sat, May 20, 2017 at 8:11 AM Tamir Duberstein <notifications@github.com>
wrote:

> Haven't looked at the main command here yet, apologies. Will finish the
> review this afternoon, but I'm concerned about the direction this is going
> with respect to baking the building binary into the docker image.
> ------------------------------
>
> Reviewed 6 of 6 files at r3, 7 of 12 files at r4.
> Review status: 7 of 12 files reviewed at latest revision, 8 unresolved
> discussions.
> ------------------------------
>
> *Makefile, line 60 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_jwEgpcBsHh07xUaw:-Kk_jwEgpcBsHh07xUax:b-c6ypik>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/Makefile#L60>):*
>
> endif
> # We intentionally use LINKFLAGS instead of the more traditional LDFLAGS
>
> this stuff used to be higher up in this file; is there a reason to
> relocate it? not relocating it has the benefit of a smaller diff and less
> git blame acquisition
> ------------------------------
>
> *Makefile, line 65 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_jul_xScyz3SeWjsT:-Kk_jul_xScyz3SeWjsU:b-7lkna3>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/Makefile#L65>):*
>
> LINKFLAGS ?=
> # TYPE=release builds a release build, which omit debug information to produce a
>
> omits
> ------------------------------
>
> *build/Dockerfile, line 137 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kdFBT8wDSAMVRk9a:-Kk_kdFBT8wDSAMVRk9b:b-fw5e6r>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/Dockerfile#L137>):*
>
> COPY cmd /root/go/src/cmdRUN GOBIN=/usr/local/bin go install ./src/cmd/... && rm -rf ./src/cmd
>
> oh boy, can we please not do this? this means we need to rev the docker
> image every time we change this command. I understand that this command
> relies on the image, but I'd rather we make the command invoke the builder
> rather than the other way around.
> ------------------------------
>
> *build/teamcity-acceptance.sh, line 15 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kXKh87fD7dGKoS4I:-Kk_kXKh87fD7dGKoS4J:b-fo4x2z>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/teamcity-acceptance.sh#L15>):*
>
> crdb-build-release
>
> do we need to install this first?
> ------------------------------
>
> *build/teamcity-build-test-binary.sh, line 5 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kUvSOJlasTdCDvZu:-Kk_kUvSOJlasTdCDvZv:b-fo4x2z>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/teamcity-build-test-binary.sh#L5>):*
>
> crdb-build-release
>
> do we need to install this first?
> ------------------------------
>
> *build/teamcity-publish-artifacts.sh, line 18 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kRrdY_BIIJl7Vo1L:-Kk_kRrdY_BIIJl7Vo1M:b-fo4x2z>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/teamcity-publish-artifacts.sh#L18>):*
>
> crdb-build-release
>
> do we need to install this first?
> ------------------------------
>
> *pkg/acceptance/prepare.sh, line 11 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-Kk_kN50mt0bzJHVZhcP:-Kk_kN50mt0bzJHVZhcQ:b72uayv>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/pkg/acceptance/prepare.sh#L11>):*
>
> crdb-build-release
>
> do we need to go install this thing first?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-:-Kk_ku7y59IMzud8NjVk:bxvxdqq>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/15337#issuecomment-302869265>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15ICnGGrAEm7Whw5GNQVAi3By19zgiks5r7thbgaJpZM4NH7Vo>
> .
>

tamird commented:
Reviewed 5 of 12 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[Makefile, line 79 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-KkbRXvO5Ks-Px_4CP2i:-KkbRXvPAmP_U0f8Z-eh:b37lt1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/Makefile#L79)):*
> ot

to

---

*[build/builder.sh, line 142 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-KkbS5rRKMmp2x8eEoGG:-KkbS5rRKMmp2x8eEoGH:b-kwb5fd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e68800f6b47c22e142b48fdb390a0e5d249a8cf/build/builder.sh#L142)):*
> ```Shell
> vols="${vols} --volume=${cockroach_toplevel}/bin.docker_amd64:/go/src/github.com/cockroachdb/cockroach/bin"
> 
> mkdir -p "${gocache}"/docker/bin
> ```

I think we need to add `gopath/pkg`, since the makefile now includes: `override GOFLAGS += -pkgdir $(GOPATH)/pkg/$(NATIVE_TAG)`; currently this doesn't seem to cache on the host at all?

---

*[build/common.mk, line 259 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-KkbQlomyClg_nU6a5RP:-KkbQlomyClg_nU6a5RQ:b-n4ip5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/build/common.mk#L259)):*
> ```Makefile
> JEMALLOC_DIR := $(BUILD_DIR)/jemalloc$(if $(USE_MSAN),_msan)
> PROTOBUF_DIR := $(BUILD_DIR)/protobuf$(if $(USE_MSAN),_msan)
> ROCKSDB_DIR  := $(BUILD_DIR)/rocksdb$(if $(USE_STDMALLOC),_stdmalloc)$(if $(USE_ROCKSDB_ASSERTIONS),_assert)
> ```

how come rocksdb is not sensitive to msan?

~~this looks to me like something that should be part of BUILD_DIR, as it before this change~~ consider adding a comment explaining why this repeated (so that protoc isn't recompiled) though perhaps is should still be solved with BUILD_DIR + TOOL_BUILD_DIR rather than repetition

---

*[pkg/cmd/publish-artifacts/main.go, line 208 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337#-KkbQ4_8-zwEPUtXCJxb:-KkbQ4_9q0AtEQYWxOgX:b-a1hud4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcbd7559a52f487c81cc48a72c56e01b94f1b8/pkg/cmd/publish-artifacts/main.go#L208)):*
> ```Go
> 					target.releaseConfig,
> 					recipe,
> 					"GOFLAGS="+extraArgs.goflags,
> ```

random style change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15337)*
<!-- Sent from Reviewable.io -->
